signingscript/docker.d/passwords.yml

```diff
@@ -12,12 +12,81 @@ in:
   $merge:
     $match:
 
-      # dep-passwords.json
-      # XXX fake-prod, aka dep, shouldn't point at autograph-external.stage
-      # for anything but autograph_stage_mar384!!
+      # We strive to:
+      # Configure `prod` environments exclusively with production Autograph entries.
+      #
+      # Configure `dev` and `fake-prod` environments with entries for production
+      # Autograph (which is what code in project repositories should be using
+      # by default), but also have entries for Autograph stage that can be
+      # opted into by tasks.
+      #
+      # The latter is to allow for easy testing of notable changes to Autograph.
+      # These entries do _not_ need to be exactly the same as the production versions,
+      # but we should be able to test each type of signature, and have a `stage_` version
+      # of each format (to make project repo changes simpler). Many of these formats
+      # will end up sharing credentials and keyids, because Autograph stage does not
+      # strive to keep a 1:1 mapping between hawkids or signers with production.
+      # This is OK; as long as we can sign with the same type of signer as production
+      # it's good enough for stage testing.
       '(ENV == "dev" || ENV == "fake-prod") && (COT_PRODUCT == "firefox" || COT_PRODUCT == "thunderbird")':
         $let:
           firefox_and_thunderbird_nonprod_autograph:
+            # GCP Autograph stage
+            - ["https://stage.autograph.nonprod.webservices.mozgcp.net",
+               "releng_authenticode_dep_sha256_2023",
+               {"$eval": "AUTOGRAPH_STAGE_AUTHENTICODE_SHA2_PASSWORD"},
```

Comment:
I'd like to understand why we're putting the autograph username in here, vs keeping it in sops along with the password.  To me that seems like it'd make rotations harder.

Reply:
This is debatable for sure (I had a thread about in Matrix, but admittedly that might not have been the right place to discuss it...).
 
The truth is that all of: username, password, and keyid are tightly coupled. There's a 1:1 relationship between username and password, and a M:1 between username and keyid. We've had issues in the past where we use the wrong username+keyid combination, and I thought it was preferable to keep those close together to help minimize it. Rotations of both username and password at the same time obviously become more annoying and error prone. Rotations of password only are maybe a little less error prone than that?

Maybe also of note is that ahal and I have been talking a bit about trying to get rid of the vast majority of username and passwords, and switching to at most 1 per trust domain, which would have access to all of the keyids that trust domain needs. At that point, password rotations are much less onerous, and arguably it makes keeping the usernames in this file more logically? (I guess you could still argue it the other way though...)

I'm not necessarily tied to keeping the usernames here, but I'm interested in your thoughts on the above tradeoffs.

Reply:
IMO rotations of password only should not be a thing ever.  They would require simultaneous deployments of autograph and signingscript, which might be acceptable in case of a really bad incident where the password got leaked and is being misused, but can just as easily be remedied by rotating both username and password.

So in my mind, username and password are tightly coupled, but keyid not so much.  If we're moving to a new keyid, we'd first configure it in autograph and deploy that, leaving credentials unchanged, then add it to signingscript and deploy it.  If we're rotating credentials, we'd add new ones to autograph, with access to the same keyids as the old user, and once it's deployed, update sops and deploy signingscript, then remove the old user from autograph (only combining the two autograph changes if the situation is bad enough that downtime is OK).

Reply:
Just to follow-up here: I'm going to move usernames back to sops. I may also make a separate patch to make sure that we log usernames + keyids by default, which will satisfy most concerns I have about ease of debugging.

Reply:
I've updated this patch to pull usernames from the env again. https://github.com/mozilla-services/cloudops-infra/pull/6093 is open for the helm part, and the `signingscript-autograph-gcp-stage-usernames` branch in the sops repo has that part prepared. Once those two are reviewed, I'll land them and re-test this patch in dev.

Reply:
Er, yes. Thank you!